### PR TITLE
Scenarios for maps with null values

### PR DESCRIPTION
**Pull request by hvub**
This PR follows up #464 with changing and augmenting TCK scenarios to reflect the behavior now clarified in the [CIP](https://github.com/sherfert/openCypher/blob/71f865baba103b54ad7bda72c360917c9c4ccc17/cip/1.accepted/CIP2015-09-16-public-type-system-type-annotation.adoc).

### CODE REVIEW

**Comment by hvub** [comment]
I will fix the red build soon. No rush.

**Comment by hvub** [comment]
Ready to be reviewed.

**Review by sherfert — APPROVED** [review]
The changes look good to me.

Before merging I would run the new scenarios with the latest neo4j.
This will make sure that we didn't make a mistake by accident
or it will uncover any bugs in the implementation earlier.

File: tck/features/expressions/comparison/Comparison1.feature [hunk 1]
```diff
@@ -143,7 +143,37 @@ Feature: Comparison1 - Equality
...
+      | {}              | {k: null}          | false  |
+      | {k: null}       | {}                 | false  |
+      | {k: 1}          | {k: 1, l: null}    | false  |
+      | {k: null, l: 1} | {l: 1}             | false  |
+      | {k: null}       | {k: null, l: null} | false  |
+      | {k: null}       | {k: null}          | null   |
+      | {k: 1}          | {k: null}          | null   |
+      | {k: 1, l: null} | {k: null, l: null} | null   |
+      | {k: 1, l: null} | {k: null, l: 1}    | null   |
+      | {k: 1, l: null} | {k: 1, l: 1}       | null   |
```
Comment (Mats-SX) [inline tck/features/expressions/comparison/Comparison1.feature:174]: 👍

File: tck/features/expressions/map/Map3.feature [hunk 2]
```diff
@@ -65,3 +65,25 @@ Feature: Map3 - Keys function
...
+      """
+      RETURN keys(<map>) AS keys
+      """
+    Then the result should be (ignoring element order for lists):
+      | keys     |
+      | <result> |
+    And no side effects
+
+    Examples:
+      | keys                  | result          |
```
Comment (Mats-SX) [inline tck/features/expressions/map/Map3.feature:81]:
```suggestion
      | map                   | result          |
```

File: tck/features/expressions/map/Map3.feature [hunk 3]
```diff
@@ -65,3 +65,25 @@ Feature: Map3 - Keys function
...
+      | {}                    | []              |
+      | {k: 1}                | ['k']           |
+      | {k: null}             | ['k']           |
+      | {null: null}          | ['null']        |
```
Comment (Mats-SX) [inline tck/features/expressions/map/Map3.feature:85]: 🙈

Comment (Mats-SX) [reply tck/features/expressions/map/Map3.feature:85]: I see the point of adding this here, but technically the key `null` is no different from the key `k`. It is amusing, but it is potentially confusing, by making an unaware reader think that Cypher supports some kind of odd concept of `null` keys... which we don't. Maybe we should not keep this?

File: tck/features/expressions/map/Map4.feature [hunk 4]
```diff
@@ -35,18 +35,28 @@ Feature: Map4 - Field existence check
...
-      | {name: 'Mats', name2: 'Pontus'} | true   |
-      | {name: null}                    | false  |
-      | {notName: 0, notName2: null}    | false  |
+      | map                                 | key     | result |
+      | {name: 'Mats', name2: 'Pontus'}     | name    | true   |
+      | {name: 'Mats', name2: 'Pontus'}     | name2   | true   |
+      | {name: null}                        | name    | true   |
+      | {name: null, name2: 'Pontus'}       | name    | true   |
+      | {name: null, name2: null}           | name    | true   |
+      | {name: null, name2: null}           | name2   | true   |
```
Comment (Mats-SX) [inline tck/features/expressions/map/Map4.feature:52]: what is the difference between these two cases? Isn't one redundant?

File: tck/features/expressions/map/Map4.feature [hunk 5]
```diff
@@ -35,18 +35,28 @@ Feature: Map4 - Field existence check
...
+      | {name: 'Pontus', name2: null}       | name2   | true   |
```
Comment (Mats-SX) [inline tck/features/expressions/map/Map4.feature:53]: this looks identical to line 50; one null value and one non-null, and then we reach for the key with null value.

File: tck/features/expressions/map/Map4.feature [hunk 6]
```diff
@@ -35,18 +35,28 @@ Feature: Map4 - Field existence check
...
+      | {name: 'Pontus', notName2: null}    | name    | true   |
+      | {notName: 'Pontus', notName2: null} | notName | true   |
```
Comment (Mats-SX) [inline tck/features/expressions/map/Map4.feature:55]: same as line 54

File: tck/features/expressions/map/Map4.feature [hunk 7]
```diff
@@ -35,18 +35,28 @@ Feature: Map4 - Field existence check
...
+      | {notName: null, notName2: null}     | notName | true   |
```
Comment (Mats-SX) [inline tck/features/expressions/map/Map4.feature:56]: same as line 51

File: tck/features/expressions/map/Map4.feature [hunk 8]
```diff
@@ -35,18 +35,28 @@ Feature: Map4 - Field existence check
...
+      | {notName: null, notName2: null}     | name    | false  |
+      | {notName: 0, notName2: null}        | name    | false  |
+      | {notName: 0}                        | name    | false  |
```
Comment (Mats-SX) [inline tck/features/expressions/map/Map4.feature:59]: I think we should also make sure to have a case with an empty map

**Review by Mats-SX — CHANGES_REQUESTED** [review]
I like the comprehensiveness of the PR. I found two bugs and some duplicated test cases. I also have some reservations around using `{null: null}` as an example case; I think it isn't helpful, and I hope we can even make it invalid.

File: tck/features/clauses/set/Set4.feature [hunk 9]
```diff
@@ -61,7 +61,26 @@ Feature: Set4 - Set All Properties with a Map
...
       """
     Then the result should be, in any order:
-      | n                           |
+      | n                          |
+      | (:X {name: 'B', baz: 'C'}) |
+    And the side effects should be:
+      | +properties | 2 |
+      | -properties | 2 |
+
+  Scenario: [2] Null values in a property map are removed with SET
```
[the remark this answers is not in the document]
Comment (hvub) [reply tck/features/clauses/set/Set4.feature:70]: Yeah, could be turned into an outline, the name would become less clear. I rather keep them separated and fix the numbering. Thanks for spotting.

File: tck/features/expressions/map/Map4.feature (hunk 4 above)
Comment (hvub) [reply tck/features/expressions/map/Map4.feature:52]: Well the second column is. Side-by-side, the two examples show, that it does not matter which field you test existence on or where this field in the order given by the literal.

File: tck/features/expressions/map/Map4.feature (hunk 5 above)
Comment (hvub) [reply tck/features/expressions/map/Map4.feature:53]: Again, the order of field does not matter. If you say, these scenarios are identical, you implicitly assume an implementation follows the rule that with field order does not matter. Go tests do not make such assumptions.

File: tck/features/expressions/map/Map4.feature (hunk 6 above)
Comment (hvub) [reply tck/features/expressions/map/Map4.feature:55]: agree

File: tck/features/expressions/map/Map4.feature (hunk 7 above)
Comment (hvub) [reply tck/features/expressions/map/Map4.feature:56]: agree

File: tck/features/expressions/map/Map4.feature (hunk 8 above)
Comment (hvub) [reply tck/features/expressions/map/Map4.feature:59]: absolutely.

File: tck/features/expressions/map/Map3.feature (hunk 3 above)
Comment (hvub) [reply tck/features/expressions/map/Map3.feature:85]: Well, where are different opinions on this topic and currently `null` is allowed, so it make sense to test that it does not confuse the keys function. It is the corner cases that are interesting for testing and blow up in the wild at some point if they have not been tested.

But since the topic of reserved keyword is still controversial, I do not mind removing the example.

**Comment by hvub** [comment]
@sherfert and @Mats-SX, thank you for your timely reviews.

@Mats-SX, I hope I have addressed all you comments sufficiently, where I agree with them. You may want to give it a second look and react to it.

File: tck/features/expressions/map/Map4.feature (hunk 5 above)
Comment (Mats-SX) [reply tck/features/expressions/map/Map4.feature:53]: I would've thought it was another test's responsibility to check that field order in maps have no meaning, but I'm happy to keep the scenario given your reasoning.

File: tck/features/expressions/map/Map3.feature (hunk 3 above)
Comment (Mats-SX) [reply tck/features/expressions/map/Map3.feature]: I mean, the only reason you wrote this is to illustrate that `null` is treated in no special way when written in a map key position; in effect it is unrelated to what this scenario is testing for. I'm not necessarily against a `MapLiteral` feature scenario including both this literal as well as `{true: false, false: true, `1`: 1}` kind of maps, but then I would say it should be clear in the scenario description that it deals with the topic of map keys.

**Review by Mats-SX — APPROVED** [review]

**Comment by Mats-SX** [comment]
@hvub Thanks for the changes -- LGTM 👍
